packages/camera/camera_android_camerax/example/integration_test/integration_test.dart

```diff
@@ -31,27 +73,94 @@ void main() {
+        expect(!(!previousPresetExactlySupported && presetExactlySupported),
+            isTrue,
+            reason:
+                'The camera took higher resolution pictures at a lower resolution.');
+        previousPresetExactlySupported = presetExactlySupported;
+        await controller.dispose();
+      }
+    }
+  });
+
+  testWidgets('Preview takes expected resolution from preset',
+      (WidgetTester tester) async {
+    final List<CameraDescription> cameras =
+        await CameraPlatform.instance.availableCameras();
+    if (cameras.isEmpty) {
+      return;
+    }
+    for (final CameraDescription cameraDescription in cameras) {
+      bool previousPresetExactlySupported = true;
+      for (final MapEntry<ResolutionPreset, Size> preset
+          in presetExpectedSizes.entries) {
+        final CameraController controller =
+            CameraController(cameraDescription, preset.key);
+
+        await controller.initialize();
+
+        while (controller.value.previewSize == null) {
+          // Wait for preview size to update.
+        }
+
+        final bool presetExactlySupported = assertExpectedDimensions(
+            presetExpectedSizes[preset.key]!, controller.value.previewSize!);
+        expect(!(!previousPresetExactlySupported && presetExactlySupported),
+            isTrue,
+            reason: 'The preview has a lower resolution than that specified.');
+        previousPresetExactlySupported = presetExactlySupported;
+        await controller.dispose();
+      }
+    }
+  });
 
-      // Take Picture
-      final XFile file = await controller.takePicture();
+  testWidgets('Images from streaming have expected resolution from preset',
+      (WidgetTester tester) async {
+    final List<CameraDescription> cameras =
+        await CameraPlatform.instance.availableCameras();
+    if (cameras.isEmpty) {
+      return;
+    }
+    for (final CameraDescription cameraDescription in cameras) {
+      bool previousPresetExactlySupported = true;
+      for (final MapEntry<ResolutionPreset, Size> preset
+          in presetExpectedSizes.entries) {
+        final CameraController controller =
+            CameraController(cameraDescription, preset.key);
+        final Completer<CameraImage> imageCompleter = Completer<CameraImage>();
+        await controller.initialize();
+        await controller.startImageStream((CameraImage image) {
+          imageCompleter.complete(image);
+          controller.stopImageStream();
+        });
 
-      // Try loading picture
-      final File fileImage = File(file.path);
-      final Image image =
-          await decodeImageFromList(fileImage.readAsBytesSync());
+        final CameraImage image = await imageCompleter.future;
+        final bool presetExactlySupported = assertExpectedDimensions(
+            presetExpectedSizes[preset.key]!,
+            Size(image.height.toDouble(), image.width.toDouble()));
+        expect(!(!previousPresetExactlySupported && presetExactlySupported),
```

Comment:
This expect is checking that (for a specific camera on the device) `presetExactlySupported` is either always false or always true (across all resolutions). Is the reason here that it should either always fallback, or always be able to exactly support? 

Can you add a comment either way explaining the purpose, as its not immediately clear to me from reading the test

Reply:
Added a comment and also re-wrote the boolean since it's overcomplicated. By the way, this is the integration test in `camera_android`, so I too had to interpret these test cases. My understanding is that if one of the lower resolution presets must fall back, then we wouldn't expect the camera to suddenly be able to reach higher resolutions; that would indicate something is wrong with our logic.

Reply:
Replaced `VideoQualityConstraint` with `VideoQuality`/`VideoQualityData` to take advantage of pigeon's support for enums + classes.